compiler/src/dotty/tools/dotc/transform/Recheck.scala

```diff
@@ -0,0 +1,329 @@
+          fntpe.instantiate(argTypes)
+
+    def recheckTypeApply(tree: TypeApply, pt: Type)(using Context): Type =
+      recheck(tree.fun).widen match
+        case fntpe: PolyType =>
+          assert(sameLength(fntpe.paramInfos, tree.args))
+          val argTypes = tree.args.map(recheck(_))
+          fntpe.instantiate(argTypes)
+
+    def recheckTyped(tree: Typed)(using Context): Type =
+      val tptType = recheck(tree.tpt)
+      recheck(tree.expr, tptType)
+      tptType
+
+    def recheckAssign(tree: Assign)(using Context): Type =
+      val lhsType = recheck(tree.lhs)
+      recheck(tree.rhs, lhsType.widen)
+      defn.UnitType
+
+    def recheckBlock(stats: List[Tree], expr: Tree, pt: Type)(using Context): Type =
+      recheckStats(stats)
+      val exprType = recheck(expr, pt.dropIfProto)
+      TypeOps.avoid(exprType, localSyms(stats).filterConserve(_.isTerm))
+
+    def recheckBlock(tree: Block, pt: Type)(using Context): Type =
+      recheckBlock(tree.stats, tree.expr, pt)
+
+    def recheckInlined(tree: Inlined, pt: Type)(using Context): Type =
+      recheckBlock(tree.bindings, tree.expansion, pt)
+
+    def recheckIf(tree: If, pt: Type)(using Context): Type =
+      recheck(tree.cond, defn.BooleanType)
+      recheck(tree.thenp, pt) | recheck(tree.elsep, pt)
+
+    def recheckClosure(tree: Closure, pt: Type)(using Context): Type =
+      if tree.tpt.isEmpty then
+        tree.meth.tpe.widen.toFunctionType(tree.meth.symbol.is(JavaDefined))
+      else
+        recheck(tree.tpt)
+
+    def recheckMatch(tree: Match, pt: Type)(using Context): Type =
+      val selectorType = recheck(tree.selector)
+      val casesTypes = tree.cases.map(recheck(_, selectorType.widen, pt))
+      TypeComparer.lub(casesTypes)
+
+    def recheck(tree: CaseDef, selType: Type, pt: Type)(using Context): Type =
+      recheck(tree.pat, selType)
+      recheck(tree.guard, defn.BooleanType)
+      recheck(tree.body, pt)
+
+    def recheckReturn(tree: Return)(using Context): Type =
+      recheck(tree.expr, tree.from.symbol.returnProto)
+      defn.NothingType
+
+    def recheckWhileDo(tree: WhileDo)(using Context): Type =
+      recheck(tree.cond, defn.BooleanType)
+      recheck(tree.body, defn.UnitType)
+      defn.UnitType
+
+    def recheckTry(tree: Try, pt: Type)(using Context): Type =
+      val bodyType = recheck(tree.expr, pt)
+      val casesTypes = tree.cases.map(recheck(_, defn.ThrowableType, pt))
+      val finalizerType = recheck(tree.finalizer, defn.UnitType)
+      TypeComparer.lub(bodyType :: casesTypes)
+
+    def recheckSeqLiteral(tree: SeqLiteral, pt: Type)(using Context): Type =
+      val elemProto = pt.stripNull.elemType match
+        case NoType => WildcardType
+        case bounds: TypeBounds => WildcardType(bounds)
+        case elemtp => elemtp
+      val declaredElemType = recheck(tree.elemtpt)
+      val elemTypes = tree.elems.map(recheck(_, elemProto))
+      TypeComparer.lub(declaredElemType :: elemTypes)
```

Comment:
Should this return the lub of the elem types or `Seq[lub of elem types]`?

Reply:
The latter. Amazingly this was not found because the checkConforms function itself did not work correctly.